numba/targets/cpu.py

```diff
@@ -186,6 +186,41 @@ def calc_array_sizeof(self, ndim):
         return self.get_abi_sizeof(self.get_value_type(aryty))
 
 
+class FastMathOptions(object):
+    """
+    Options for controlling fast math optimization.
+    """
+    def __init__(self, value):
+        # https://releases.llvm.org/7.0.0/docs/LangRef.html#fast-math-flags
+        valid_flags = {
+            'fast',
+            'nnan', 'ninf', 'nsz', 'arcp',
+            'contract', 'afn', 'reassoc',
+        }
+
+        if value is True:
```

Comment:
Should it be an error if:
1. Same flag is passed twice (llvm itself doesn't care, e.g. `fmul nnan nnan nnan double %1, %2` is fine).
2. Same flag is passed twice with differing value.
3. `fast` is present and any other flag is present (llvm itself doesn't care, e.g. `fmul fast nnan double %1, %2` is fine).

?

Reply:
RE points 1 and 2 I'm not sure how it would be possible to get the same option twice, because I only accept `dict` and `set`.

RE point 3 this is more arguably, but my instinct is to err on the side of not erroring out. You can imagine that a user might e.g. build a fastmath flag like `@njit(fastmath={'nnan'} | ({'fast'} if foo else set()))`, and I think it would be surprising if this was an error if `foo` were true because the intent of the call is clearly to enable all optimisations if `foo` is set.

Reply:
RE 1 & 2. Oh for some reason I thought list was in there too, apologies.

RE 3. I think this is reasonable until determined otherwise.